File: agenttel-go/enums/enums.go
// Package enums defines typed string constants for all AgentTel domain enumerations.
package enums

// ServiceTier indicates the criticality tier of a service.
type ServiceTier string

const (
	ServiceTierCritical     ServiceTier = "critical"
	ServiceTierStandard     ServiceTier = "standard"
	ServiceTierInternal     ServiceTier = "internal"
	ServiceTierExperimental ServiceTier = "experimental"
)

// ErrorCategory classifies errors for automated triage.
type ErrorCategory string

const (
	ErrorCategoryDependencyTimeout  ErrorCategory = "dependency_timeout"
	ErrorCategoryConnectionError    ErrorCategory = "connection_error"
	ErrorCategoryCodeBug            ErrorCategory = "code_bug"
	ErrorCategoryRateLimited        ErrorCategory = "rate_limited"
	ErrorCategoryAuthFailure        ErrorCategory = "auth_failure"
	ErrorCategoryResourceExhaustion ErrorCategory = "resource_exhaustion"
	ErrorCategoryDataValidation     ErrorCategory = "data_validation"
	ErrorCategoryUnknown            ErrorCategory = "unknown"
)

// DependencyType identifies the kind of external dependency.
type DependencyType string

const (
	DependencyTypeInternalService  DependencyType = "internal_service"
	DependencyTypeExternalAPI      DependencyType = "external_api"
	DependencyTypeDatabase         DependencyType = "database"
	DependencyTypeMessageBroker    DependencyType = "message_broker"
	DependencyTypeCache            DependencyType = "cache"
	DependencyTypeObjectStore      DependencyType = "object_store"
	DependencyTypeIdentityProvider DependencyType = "identity_provider"
)

// DependencyCriticality indicates how critical a dependency is.
type DependencyCriticality string

const (
	DependencyCriticalityRequired DependencyCriticality = "required"
	DependencyCriticalityDegraded DependencyCriticality = "degraded"
	DependencyCriticalityOptional DependencyCriticality = "optional"
)

// DependencyState indicates the health state of a dependency.
type DependencyState string

const (
	DependencyStateHealthy   DependencyState = "healthy"
	DependencyStateDegraded  DependencyState = "degraded"
	DependencyStateUnhealthy DependencyState = "unhealthy"
	DependencyStateUnknown   DependencyState = "unknown"
)

// BaselineSource indicates the origin of baseline data.
type BaselineSource string

const (
	BaselineSourceStatic    BaselineSource = "static"
	BaselineSourceRolling7D BaselineSource = "rolling_7d"
	BaselineSourceMLModel   BaselineSource = "ml_model"
	BaselineSourceSLO       BaselineSource = "slo"
)

// EscalationLevel indicates the severity of escalation.
type EscalationLevel string

const (
	EscalationLevelAutoResolve       EscalationLevel = "auto_resolve"
	EscalationLevelNotifyTeam        EscalationLevel = "notify_team"
	EscalationLevelPageOnCall        EscalationLevel = "page_oncall"
	EscalationLevelIncidentCommander EscalationLevel = "incident_commander"
)

// ConsumptionPattern describes how a consumer calls a service.
type ConsumptionPattern string

const (
	ConsumptionPatternSync     ConsumptionPattern = "sync"
	ConsumptionPatternAsync    ConsumptionPattern = "async"
	ConsumptionPatternBatch    ConsumptionPattern = "batch"
	ConsumptionPatternStreaming ConsumptionPattern = "streaming"
)

// IncidentPattern identifies known failure patterns detectable from telemetry.
type IncidentPattern string

const (
	IncidentPatternCascadeFailure     IncidentPattern = "cascade_failure"
	IncidentPatternMemoryLeak         IncidentPattern = "memory_leak"
	IncidentPatternThunderingHerd     IncidentPattern = "thundering_herd"
	IncidentPatternColdStart          IncidentPattern = "cold_start"
	IncidentPatternErrorRateSpike     IncidentPattern = "error_rate_spike"
	IncidentPatternLatencyDegradation IncidentPattern = "latency_degradation"
)

// SLOType identifies the type of service level objective.
type SLOType string

const (
	SLOTypeAvailability SLOType = "availability"
	SLOTypeLatencyP99   SLOType = "latency_p99"
	SLOTypeLatencyP50   SLOType = "latency_p50"
	SLOTypeErrorRate    SLOType = "error_rate"
)

// AlertSeverity indicates the severity of an SLO alert.
type AlertSeverity string

const (
	AlertSeverityInfo     AlertSeverity = "info"
	AlertSeverityWarning  AlertSeverity = "warning"
	AlertSeverityCritical AlertSeverity = "critical"
)

// AgentType indicates the role of an AI agent in an orchestration.
type AgentType string

const (
	AgentTypeSingle       AgentType = "single"
	AgentTypeOrchestrator AgentType = "orchestrator"
	AgentTypeWorker       AgentType = "worker"
	AgentTypeEvaluator    AgentType = "evaluator"
	AgentTypeCritic       AgentType = "critic"
	AgentTypeRouter       AgentType = "router"
)

// StepType identifies the type of reasoning step.
type StepType string

const (
	StepTypeThought     StepType = "thought"
	StepTypeAction      StepType = "action"
	StepTypeObservation StepType = "observation"
	StepTypeEvaluation  StepType = "evaluation"
	StepTypeRevision    StepType = "revision"
)

// OrchestrationPattern identifies multi-agent coordination patterns.
type OrchestrationPattern string

const (
	OrchestrationPatternReact              OrchestrationPattern = "react"
	OrchestrationPatternSequential         OrchestrationPattern = "sequential"
	OrchestrationPatternParallel           OrchestrationPattern = "parallel"
	OrchestrationPatternHandoff            OrchestrationPattern = "handoff"
	OrchestrationPatternOrchestratorWorkers OrchestrationPattern = "orchestrator_workers"
	OrchestrationPatternEvaluatorOptimizer OrchestrationPattern = "evaluator_optimizer"
	OrchestrationPatternGroupChat          OrchestrationPattern = "group_chat"
	OrchestrationPatternSwarm              OrchestrationPattern = "swarm"
	OrchestrationPatternHierarchical       OrchestrationPattern = "hierarchical"
)

// InvocationStatus indicates the outcome of an agent invocation.
type InvocationStatus string

const (
	InvocationStatusRunning   InvocationStatus = "running"
	InvocationStatusCompleted InvocationStatus = "completed"
	InvocationStatusFailed    InvocationStatus = "failed"
	InvocationStatusCancelled InvocationStatus = "cancelled"
)

// GuardrailAction indicates what a guardrail did when triggered.
type GuardrailAction string

const (
	GuardrailActionBlock    GuardrailAction = "block"
	GuardrailActionWarn     GuardrailAction = "warn"
	GuardrailActionModify   GuardrailAction = "modify"
	GuardrailActionEscalate GuardrailAction = "escalate"
)

// MemoryOperation indicates the type of agent memory operation.
type MemoryOperation string

const (
	MemoryOperationRead   MemoryOperation = "read"
	MemoryOperationWrite  MemoryOperation = "write"
	MemoryOperationDelete MemoryOperation = "delete"
	MemoryOperationSearch MemoryOperation = "search"
)

// HumanCheckpointType indicates the type of human-in-the-loop checkpoint.
type HumanCheckpointType string

const (
	HumanCheckpointTypeApproval HumanCheckpointType = "approval"
	HumanCheckpointTypeReview   HumanCheckpointType = "review"
	HumanCheckpointTypeInput    HumanCheckpointType = "input"
	HumanCheckpointTypeOverride HumanCheckpointType = "override"
)

// EvalType indicates the type of evaluation scorer.
type EvalType string

const (
	EvalTypeLLMJudge EvalType = "llm_judge"
	EvalTypeRule     EvalType = "rule"
	EvalTypeHuman    EvalType = "human"
	EvalTypeCustom   EvalType = "custom"
)

// ErrorSource identifies where an agentic error originated.
type ErrorSource string

const (
	ErrorSourceLLM    ErrorSource = "llm"
	ErrorSourceTool   ErrorSource = "tool"
	ErrorSourceCode   ErrorSource = "code"
	ErrorSourceInput  ErrorSource = "input"
	ErrorSourceSystem ErrorSource = "system"
)

File: agenttel-types/src/enums.ts
/** Service criticality tier. */
export enum ServiceTier {
  CRITICAL = 'critical',
  STANDARD = 'standard',
  INTERNAL = 'internal',
  EXPERIMENTAL = 'experimental',
}

/** Error classification categories for automated triage. */
export enum ErrorCategory {
  DEPENDENCY_TIMEOUT = 'dependency_timeout',
  CONNECTION_ERROR = 'connection_error',
  CODE_BUG = 'code_bug',
  RATE_LIMITED = 'rate_limited',
  AUTH_FAILURE = 'auth_failure',
  RESOURCE_EXHAUSTION = 'resource_exhaustion',
  DATA_VALIDATION = 'data_validation',
  UNKNOWN = 'unknown',
}

/** Type of external dependency. */
export enum DependencyType {
  INTERNAL_SERVICE = 'internal_service',
  EXTERNAL_API = 'external_api',
  DATABASE = 'database',
  MESSAGE_BROKER = 'message_broker',
  CACHE = 'cache',
  OBJECT_STORE = 'object_store',
  IDENTITY_PROVIDER = 'identity_provider',
}

/** How critical a dependency is. */
export enum DependencyCriticality {
  REQUIRED = 'required',
  DEGRADED = 'degraded',
  OPTIONAL = 'optional',
}

/** Health state of a dependency. */
export enum DependencyState {
  HEALTHY = 'healthy',
  DEGRADED = 'degraded',
  UNHEALTHY = 'unhealthy',
  UNKNOWN = 'unknown',
}

/** Source of baseline data. */
export enum BaselineSource {
  STATIC = 'static',
  ROLLING_7D = 'rolling_7d',
  ML_MODEL = 'ml_model',
  SLO = 'slo',
}

/** Escalation severity level. */
export enum EscalationLevel {
  AUTO_RESOLVE = 'auto_resolve',
  NOTIFY_TEAM = 'notify_team',
  PAGE_ONCALL = 'page_oncall',
  INCIDENT_COMMANDER = 'incident_commander',
}

/** How a consumer calls a service. */
export enum ConsumptionPattern {
  SYNC = 'sync',
  ASYNC = 'async',
  BATCH = 'batch',
  STREAMING = 'streaming',
}

/** Known incident failure patterns. */
export enum IncidentPattern {
  CASCADE_FAILURE = 'cascade_failure',
  MEMORY_LEAK = 'memory_leak',
  THUNDERING_HERD = 'thundering_herd',
  COLD_START = 'cold_start',
  ERROR_RATE_SPIKE = 'error_rate_spike',
  LATENCY_DEGRADATION = 'latency_degradation',
}

/** Type of service level objective. */
export enum SLOType {
  AVAILABILITY = 'availability',
  LATENCY_P99 = 'latency_p99',
  LATENCY_P50 = 'latency_p50',
  ERROR_RATE = 'error_rate',
}

/** Severity of an SLO alert. */
export enum AlertSeverity {
  INFO = 'info',
  WARNING = 'warning',
  CRITICAL = 'critical',
}

/** Role of an AI agent in an orchestration. */
export enum AgentType {
  SINGLE = 'single',
  ORCHESTRATOR = 'orchestrator',
  WORKER = 'worker',
  EVALUATOR = 'evaluator',
  CRITIC = 'critic',
  ROUTER = 'router',
}

/** Type of reasoning step. */
export enum StepType {
  THOUGHT = 'thought',
  ACTION = 'action',
  OBSERVATION = 'observation',
  EVALUATION = 'evaluation',
  REVISION = 'revision',
}

/** Multi-agent coordination patterns. */
export enum OrchestrationPattern {
  REACT = 'react',
  SEQUENTIAL = 'sequential',
  PARALLEL = 'parallel',
  HANDOFF = 'handoff',
  ORCHESTRATOR_WORKERS = 'orchestrator_workers',
  EVALUATOR_OPTIMIZER = 'evaluator_optimizer',
  GROUP_CHAT = 'group_chat',
  SWARM = 'swarm',
  HIERARCHICAL = 'hierarchical',
}

/** Outcome of an agent invocation. */
export enum InvocationStatus {
  RUNNING = 'running',
  COMPLETED = 'completed',
  FAILED = 'failed',
  CANCELLED = 'cancelled',
}

/** Guardrail action when triggered. */
export enum GuardrailAction {
  BLOCK = 'block',
  WARN = 'warn',
  MODIFY = 'modify',
  ESCALATE = 'escalate',
}

/** Agent memory operation type. */
export enum MemoryOperation {
  READ = 'read',
  WRITE = 'write',
  DELETE = 'delete',
  SEARCH = 'search',
}

/** Human-in-the-loop checkpoint type. */
export enum HumanCheckpointType {
  APPROVAL = 'approval',
  REVIEW = 'review',
  INPUT = 'input',
  OVERRIDE = 'override',
}

/** Evaluation scorer type. */
export enum EvalType {
  LLM_JUDGE = 'llm_judge',
  RULE = 'rule',
  HUMAN = 'human',
  CUSTOM = 'custom',
}

/** Where an agentic error originated. */
export enum ErrorSource {
  LLM = 'llm',
  TOOL = 'tool',
  CODE = 'code',
  INPUT = 'input',
  SYSTEM = 'system',
}
